### PR TITLE
Fix #17053: Crash when trying to open files under 4 bytes in length

### LINKED ISSUE #17053: [774595] openrct2.exe: sawyercoding_detect_file_type, TryClassifyAsS4, TryClassifyFile, OpenRCT2::Context::LoadParkFromStream...
<table>
    <tr>
        <td></td>
        <td><b><a href="https://openrct2.sp.backtrace.io/dashboard/openrct2/project/openrct2/group/774595a1f7ce4655347759cdcf4803f61657ead3ac38339dbca49f592d31fd56">New error: [openrct2]</a></b><br /><b>1</b> occurrence since (Fri Nov 15 2019 11:27:39)<br /></td>
    </tr>
</table><table>
    <tr><td><strong>Application</strong></td><td>openrct2.exe</td></tr>
<tr>
    <td><strong>Attributes</strong></td>
    <td>
<strong>_rxid</strong> &mdash; 1
        <ul>
                <li><code>0a000000-0e5d-1608-0000-000000000000</code> &mdash; 1</li></ul><br>
<strong>attachment_config.ini</strong> &mdash; 1
        <ul>
                <li><code>C:\Users\ankhh\Documents\OpenRCT2\crash\37782b19-c8ac-4375-b9c0-f24ea366ad73.ini</code> &mdash; 1</li></ul><br>
<strong>assert_failure</strong> &mdash; 1
        <ul>
                <li><code>*</code> &mdash; 1</li></ul><br>
<strong>fault.address</strong> &mdash; 1
        <ul>
                <li><code>273072128</code> &mdash; 1</li></ul><br>
<strong>version</strong> &mdash; 1
        <ul>
                <li><code>OpenRCT2, v0.3.5.1-906-gdaeaa98 (daeaa98 on develop) provided by GitHub</code> &mdash; 1</li></ul><br>
<strong>error.message</strong> &mdash; 1
        <ul>
                <li><code>EXCEPTION_ACCESS_VIOLATION_READ</code> &mdash; 1</li></ul><br>
<strong>commit</strong> &mdash; 1
        <ul>
                <li><code>daeaa98</code> &mdash; 1</li></ul><br></td>
    </tr>
    <tr>
        <td><strong>Classifiers</strong></td>
        <td>
            <strong>invalid-read</strong><br></td>
    </tr>
    <tr>
        <td><strong>Frames</strong></td>
        <td>
            <code>
sawyercoding_detect_file_type
TryClassifyAsS4
TryClassifyFile
OpenRCT2::Context::LoadParkFromStream
OpenRCT2::Context::LoadParkFromFile
game_load_or_quit_no_save_prompt_callback
std::_Func_impl_no_alloc&lt;&lt;lambda_bcf1c7815abc4622d7c3991f47e8186f&gt;,void,int,std::basic_string_view&lt;char,std::char_traits&lt;char&gt; &gt; &gt;::_Do_call
WindowLoadsaveInvokeCallback
WindowLoadsaveSelect</code>
        </td>
    </tr></table>

### PR DESCRIPTION
**Pull request by Gymnasiast**
This is responsible for a *lot* of crash reports.

### CODE REVIEW

File: src/openrct2/util/SawyerCoding.cpp [hunk 1]
```diff
@@ -384,6 +387,11 @@ static void encode_chunk_rotate(uint8_t* buffer, size_t length)
 
 int32_t sawyercoding_detect_file_type(const uint8_t* src, size_t length)
 {
+    if (length < 4)
+    {
+        throw std::runtime_error("Stream is (nearly) empty!");
```
Comment (Broxzier) [inline src/openrct2/util/SawyerCoding.cpp:392]: `std::runtime_error` is for unpredictable events, while the case is very clear here. `std::length_error` seems more appropriate here.

```suggestion
        throw std::length_error("Stream is (nearly) empty!");
```

Comment (Broxzier) [reply src/openrct2/util/SawyerCoding.cpp:392]: Also don't forget to include <stdexcept> in this source file.

**Comment by Gymnasiast** [comment]
@Broxzier Done!

**Review by Broxzier — APPROVED** [review]
Changes look sane, and I assume you have verified this fix yourself.

**Comment by Gymnasiast** [comment]
I have, yes, with files of differing sizes (which crashes OpenRCT2 before the fix).

**Comment by Gymnasiast** [comment]
> How does that happen (or how does one fixes this if it happens for them)?
> 
> I'm clueless

Like I said, just try opening a file under 4 bytes in length. The function simply assumed that the file was always going to be at least that big, and made assumptions accordingly in their use of array offsets.

**Comment by IntelOrca** [comment]
I suppose the question is, why is it so common? Who keeps trying to load < 4 byte files?

**Comment by Gymnasiast** [comment]
I guess corrupted files could be an explanation.

**Comment by tupaschoal** [comment]
> I suppose the question is, why is it so common? Who keeps trying to load < 4 byte files?

Yep, that's what I actually wanted to understand

**Pull request merged**